### PR TITLE
FIX: Set evaluation mode to compute number of dimensions

### PR DESCRIPTION
**Pull request by fepegar**
When guessing the number of output dimensions in the image encoder, the model is in training mode and the batch norm stats get updated. This shouldn't really happen and makes the checks in https://github.com/microsoft/hi-ml/pull/807/files#r1112200449 fail. This PR adds a line to ensure the encode is in `eval` mode before calculating the number of output dims.

### CODE REVIEW

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ def __init__(self,
 
         # Initiate encoder, projector, and classifier
         self.encoder = ImageEncoder(img_model_type)
+        self.encoder.eval()  # to get the output size in the next line
```
Comment (ozan-oktay) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:90]: By convention `nn.modules` are initialised with `self.training` set to `True`.

Instead of hardcoding this here, perhaps we could handle this inside the `get_encoder_output_dim` method also in case other classes use this function later.

One idea is that -- In line 214, you could add the following:
```
        initial_state = self.encoder.training
        module.eval()
        module.forward(...)
        module.train(initial_state) # Restore the state of encoder again
```

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:90]: Thanks, I thought of doing that, potentially with a context manager. But I think in this case it shouldn't be necessary as we have `self.train()` in the `__init__` of the class, just a few lines below: https://github.com/microsoft/hi-ml/pull/809/files#diff-ddbcf0079364b4612521f029a11af075c8293ec9c06f147e9cedafa17f6c42bdR99

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:91]: I've created a context manager in [4bdc2e7](https://github.com/microsoft/hi-ml/pull/809/commits/4bdc2e7bf8cd61bb53431c827955679229e21b4c) to handle this in a generic way.

Comment (Shruthi42) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:91]: Looks good to me, if we remove this line and use the context manager you created instead.

**Review by Shruthi42 — APPROVED** [review]

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py [hunk 2]
```diff
@@ -212,5 +214,18 @@ def get_encoder_output_dim(module: torch.nn.Module) -> int:
...
     # Extract the number of output feature dimensions
-    representations = module(x)
+    with restore_training_mode(module):
+        module.eval()
+        representations = module(x)
     return representations.shape[1]
+
+
+@contextmanager
+def restore_training_mode(module: nn.Module) -> Generator[None, None, None]:
```
Comment (Shruthi42) [inline hi-ml-multimodal/src/health_multimodal/image/model/model.py:224]: @fepegar This method could be useful in other codebases. Wondering if it should be placed elsewhere, instead of in image/model.

Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:224]: It should, I agree. But the import from `hubconf.py` is a bit brittle as it's designed so only `torch` and `torchvision` are needed, so I'd rather work on that on a different PR so we can unblock @markpinnock's PR (https://github.com/microsoft/hi-ml/pull/807). Does that make sense?

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 1 above)
Comment (fepegar) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py]: Thanks! I forgot to remove it. Done in 9533bead.

**Review by ozan-oktay — APPROVED** [review]
👍

File: hi-ml-multimodal/src/health_multimodal/image/model/model.py (hunk 2 above)
Comment (Shruthi42) [reply hi-ml-multimodal/src/health_multimodal/image/model/model.py:223]: I see, this sounds good.
